### PR TITLE
radio button implementation

### PR DESCRIPTION
**Pull request by DhaaraniCIT**

### CODE REVIEW

**Comment by DhaaraniCIT** [comment]
Screenshot 2023-02-02 at 4 11 46 PM

**Comment by DhaaraniCIT** [comment]
https://user-images.githubusercontent.com/55541808/216305003-30ed82fa-3b76-4b5f-aee4-a8160202eb48.mov

File: .eslintrc.json [hunk 1]
```diff
@@ -58,7 +58,7 @@
         "@angular-eslint/no-conflicting-lifecycle": "off",
         "lines-between-class-members": ["error", "always"],
         "@typescript-eslint/no-shadow": "off",
-        "complexity": [2, 10],
+        "complexity": [2, 15],
```
Comment (ashwin1111) [inline .eslintrc.json:61]: let's not increase this. Keep it to 10. Refactor complex functions to smaller chunks

File: karma.conf.js [hunk 2]
```diff
@@ -36,7 +36,7 @@ module.exports = function (config) {
       check: {
         global: {
           statements: 90,
-          branches: 85,
+          branches: 80,
```
Comment (ashwin1111) [inline karma.conf.js:39]: add a todo to increase coverage over the project completion

File: src/app/core/models/qbd/qbd-configuration/export-setting.model.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import { FormGroup } from "@angular/forms";
+import { QBDCorporateCreditCardExpensesObject, QBDEntity, QBDExpenseGroupedBy, QBDExpenseState, QBDExportDateType, QBDReimbursableExpensesObject } from "../../enum/enum.model";
```
Comment (ashwin1111) [inline src/app/core/models/qbd/qbd-configuration/export-setting.model.ts:2]: PR title and implementation doesn't match, please to update title

File: src/app/core/services/qbd/qbd-core/qbd-workspace.service.ts [hunk 4]
```diff
@@ -9,14 +12,28 @@ import { QbdApiService } from './qbd-api.service';
...
+  getWorkspaceId(): string {
+    return this.storageService.get('workspaceId');
+  }
+
+  setOnboardingState(onboardingState: QBDOnboardingState): void {
+    return this.storageService.set('onboardingState', onboardingState);
+  }
+
+  getOnboardingState(): QBDOnboardingState {
+    return this.storageService.get('onboardingState');
```
Comment (ashwin1111) [inline src/app/core/services/qbd/qbd-core/qbd-workspace.service.ts:37]: this will return null incase of no data. In that case, the return type becomes incorrect.
```
const onboardingState = this.storageService.get('onboardingState');
return onboardingState ? onboardingState : 'ONBOARDING<>OR<>ANY RELEVANT STATE' ;
```

File: src/app/integrations/integrations.module.ts [hunk 5]
```diff
@@ -17,6 +18,7 @@ import { QbdComponent } from './qbd/qbd.component';
   imports: [
     CommonModule,
     IntegrationsRoutingModule,
+    SharedModule,
```
Comment (ashwin1111) [inline src/app/integrations/integrations.module.ts:21]: do we use anything from shared module in integrations module?

File: src/app/integrations/qbd/qbd.component.html [hunk 6]
```diff
@@ -1 +1,4 @@
-<router-outlet></router-outlet>
\ No newline at end of file
+<router-outlet></router-outlet>
+<div *ngIf="isLoading" class="tw-flex tw-justify-center tw-items-center tw-h-screen">
+    <app-loader></app-loader>
+  </div>
```
Comment (ashwin1111) [inline src/app/integrations/qbd/qbd.component.html:4]: indent seems to be off

File: src/app/integrations/qbd/qbd.component.ts [hunk 7]
```diff
@@ -47,28 +48,32 @@ export class QbdComponent implements OnInit {
     }
   }
 
-  getOrCreateWorkspace(): Promise<Workspace> {
-    return this.workspaceService.getQBDWorkspace(this.user.org_id).toPromise().then((workspaces) => {
+  getOrCreateWorkspace(): void {
```
Comment (ashwin1111) [inline src/app/integrations/qbd/qbd.component.ts:51]: make it a private function

File: src/app/integrations/qbd/qbd.component.ts [hunk 8]
```diff
@@ -47,28 +48,32 @@ export class QbdComponent implements OnInit {
...
+    this.workspaceService.getQBDWorkspace(this.user.org_id).subscribe((workspaces) => {
       if (workspaces?.id) {
-        return workspaces;
+        this.workspace = workspaces;
+        this.storageService.set('workspaceId', this.workspace.id);
```
Comment (ashwin1111) [inline src/app/integrations/qbd/qbd.component.ts:55]: duplicate code, separate out reusable parts. Checkout qbo/xero

File: src/app/shared/components/qbd/configuration/configuration-radio-field/configuration-radio-field.component.scss [hunk 9]
```diff
@@ -0,0 +1,9 @@
+:host ::ng-deep .p-radiobutton
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/configuration/configuration-radio-field/configuration-radio-field.component.scss:1]: can add it globally so that even the future integrations can consume same stylings.

File: src/app/shared/components/qbd/configuration/configuration-radio-field/configuration-radio-field.component.html [hunk 10]
```diff
@@ -1 +1,17 @@
-<p>configuration-radio-field works!</p>
+<div [formGroup]="form" class="tw-p-24-px tw-flex tw-flex-row tw-space-between">
+    <div class="tw-w-3/5 tw-flex">
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/configuration/configuration-radio-field/configuration-radio-field.component.html:2]: let's use px please instead of 3/5

Comment (ashwin1111) [reply src/app/shared/components/qbd/configuration/configuration-radio-field/configuration-radio-field.component.html:2]: if the particular size is not available, you can add it to the relevant object in `tailwind.config.js`

File: src/app/shared/components/qbd/configuration/configuration-radio-field/configuration-radio-field.component.ts [hunk 11]
```diff
@@ -9,7 +11,22 @@ export class ConfigurationRadioFieldComponent implements OnInit {
 
   constructor() { }
 
+  @Input() options: any;
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/configuration/configuration-radio-field/configuration-radio-field.component.ts:14]: declare a type instead of any

File: src/app/shared/components/qbd/configuration/configuration-toggle-field/configuration-toggle-field.component.ts [hunk 12]
```diff
@@ -1,5 +1,5 @@
 import { Component, Input, OnInit } from '@angular/core';
-import { FormGroup } from '@angular/forms';
+import { FormControlName, FormGroup } from '@angular/forms';
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/configuration/configuration-toggle-field/configuration-toggle-field.component.ts:2]: remove FormControlName since this is unused

File: src/app/shared/components/qbd/configuration/export-setting/export-setting.component.html [hunk 13]
```diff
@@ -1,29 +1,103 @@
...
+                            [label]="'How should the expenses be exported?'"
+                            [subLabel]="'Choose the type of transaction in Quickbooks Desktop to export your Fyle expenses.'"
+                            [options]="reimbursableExportTypes"
+                            [iconPath]="'../../../../../../assets/icons/expense.svg'"
+                            [formControllerName]="'reimbursableExportType'"></app-configuration-radio-field>
+
+                            <div *ngIf="exportSettingsForm.value.reimbursableExportType">
+                                <div class="tw-pl-16 tw-pt-30-px tw-pr-24-px tw-pb-24-px tw-flex tw-flex-row tw-space-between">
+                                    <div class="tw-w-3/5">
+                                        <h4 class="tw-text-14-px">To which {{ exportSettingsForm.value.reimbursableExportType === QBDReimbursableExpensesObject.BILL ? 'Accounts Payable' : 'Bank'}} account should the {{ exportSettingsForm.value.reimbursableExportType === QBDReimbursableExpensesObject.BILL ? reimbursableExportTypes[0].label : reimbursableExportTypes[1].label }} be posted to?</h4>
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/configuration/export-setting/export-setting.component.html:34]:
```
{{ exportSettingsForm.value.reimbursableExportType === QBDReimbursableExpensesObject.BILL ? 'Accounts Payable' : 'Bank'}} account should the {{ exportSettingsForm.value.reimbursableExportType === QBDReimbursableExpensesObject.BILL ? reimbursableExportTypes[0].label : reimbursableExportTypes[1].label }}
```
move this logic to ts, looks more complicated to be in html

File: src/app/shared/components/qbd/configuration/export-setting/export-setting.component.html [hunk 14]
```diff
@@ -1,29 +1,103 @@
...
+                                    <div class="tw-pl-24-px">
+                                        <input type="text" class="tw-w-80 tw-px-10-px tw-py-8-px tw-border tw-border-solid !tw-border-separator tw-rounded-4-px" pInputText formControlName="bankAccount" placeholder="Enter {{ exportSettingsForm.value.reimbursableExportType === QBDReimbursableExpensesObject.BILL ? 'Accounts Payable' : 'Bank'}} Name"> 
+                                    </div>
+                                </div>
+                            </div>
+                        </div>
+                        <div class="tw-mt-12-px tw-bg-white tw-border tw-border-solid tw-border-separator tw-rounded-12-px">
+                            <app-configuration-radio-field
+                            [form]="exportSettingsForm"
+                            [label]="'How would you like to group your expenses? *'"
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/configuration/export-setting/export-setting.component.html:47]: * should be from a separate component, and the colour should be different from label

File: src/app/shared/components/qbd/configuration/export-setting/export-setting.component.html [hunk 15]
```diff
@@ -1,29 +1,103 @@
...
+                            [subLabel]="'Grouping reflects how the expense entries are posted in QBD.'"
+                            [options]="expenseGroupingFieldOptions"
+                            [iconPath]="'../../../../../../assets/icons/copyV2.svg'"
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/configuration/export-setting/export-setting.component.html:50]: copyV2 naming convention isn't right. Also use `src/assets/...` as path instead of `../../`

File: src/app/shared/components/qbd/configuration/export-setting/export-setting.component.html [hunk 16]
```diff
@@ -1,29 +1,103 @@
...
+                                </div>
+                            </div>
+                        </div>
+                        <div class="tw-mt-12-px tw-bg-white tw-border tw-border-solid tw-border-separator tw-rounded-12-px">
+                            <app-configuration-radio-field
+                            [form]="exportSettingsForm"
+                            [label]="'How would you like to group your expenses?'"
+                            [subLabel]="'Grouping reflects how the expense entries are posted in QBD.'"
+                            [options]="expenseGroupingFieldOptions"
+                            [iconPath]="'../../../../../../assets/icons/copyV2.svg'"
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/configuration/export-setting/export-setting.component.html:92]: same as above

File: src/app/shared/components/qbd/configuration/export-setting/export-setting.component.ts [hunk 17]
```diff
@@ -20,16 +25,214 @@ export class ExportSettingComponent implements OnInit {
...
+          cccExportGroup: [null],
+          cccExportDate: [null],
+          bankAccount: [null],
+          cccEntityName: [null],
+          cccAccountName: [null],
+          reimbursableExpenseState: [null],
+          cccExpenseState: [null]
+        });
+        this.setCustomValidatorsAndWatchers();
+      this.isLoading = false;
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/configuration/export-setting/export-setting.component.ts:210]: indent is off

File: src/app/shared/components/qbd/configuration/export-setting/export-setting.component.ts [hunk 18]
```diff
@@ -1,12 +1,17 @@
...
+import { ExportSettingModel, QBDExportSettingFormOption, QBDExportSettingGet } from 'src/app/core/models/qbd/qbd-configuration/export-setting.model';
+import { QbdExportSettingService } from 'src/app/core/services/qbd/qbd-configuration/qbd-export-setting.service';
+import { QbdWorkspaceService } from 'src/app/core/services/qbd/qbd-core/qbd-workspace.service';
 
 @Component({
   selector: 'app-export-setting',
   templateUrl: './export-setting.component.html',
-  styleUrls: ['./export-setting.component.scss']
+  styleUrls: ['./export-setting.component.scss'],
+  providers: [MessageService]
```
Comment (ashwin1111) [inline src/app/shared/components/qbd/configuration/export-setting/export-setting.component.ts:14]: this is not required, can add it to the module directly. no need of adding in individual components

File: src/app/shared/shared.module.ts [hunk 19]
```diff
@@ -3,6 +3,10 @@ import { CommonModule } from '@angular/common';
 import { LoaderComponent } from './components/core/loader/loader.component';
 import { AppHeaderComponent } from './components/apps/app-header/app-header.component';
 import { InputSwitchModule } from 'primeng/inputswitch';
+import { RadioButtonModule } from 'primeng/radiobutton';
+import { ToastModule } from 'primeng/toast';
+import { RippleModule } from 'primeng/ripple';
```
Comment (ashwin1111) [inline src/app/shared/shared.module.ts:8]: why is this needed?

File: src/app/shared/shared.module.ts [hunk 20]
```diff
@@ -3,6 +3,10 @@ import { CommonModule } from '@angular/common';
...
+import { InputTextModule } from 'primeng/inputtext';
```
Comment (ashwin1111) [inline src/app/shared/shared.module.ts:9]: do we have input text box somewhere?

File: src/styles.scss [hunk 21]
```diff
@@ -46,7 +46,7 @@ p {
     @apply tw-text-14-px tw-border-box-color #{!important};
 }
 
-.p-inputtext:enabled:focus {
+.p-inputtext:enabled:focus-visible {
```
Comment (ashwin1111) [inline src/styles.scss:49]: make sure this doesn't break existing integration. Check bamboo hr connection dialog if there is any impact

File: src/styles.scss [hunk 22]
```diff
@@ -92,6 +92,20 @@ p {
     @apply tw-shadow-none #{!important};
 }
 
+// Radio Button
+
```
Comment (ashwin1111) [inline src/styles.scss:96]: line break after comment is not required

File: src/test.ts [hunk 23]
```diff
@@ -21,6 +21,7 @@ getTestBed().initTestEnvironment(
 );
 
 // Then we find all the tests.
+localStorage.setItem('workspaceId', '1');
```
Comment (ashwin1111) [inline src/test.ts:24]: mock storage / workspace service in tests. No need of adding this here.

File: src/app/integrations/integrations.module.ts (hunk 5 above)
Comment (DhaaraniCIT) [reply src/app/integrations/integrations.module.ts:21]: if It is absent in integrations, app-loader in the qbd component is not working

File: src/app/shared/shared.module.ts (hunk 20 above)
Comment (DhaaraniCIT) [reply src/app/shared/shared.module.ts:9]: yes

File: src/app/shared/components/qbd/configuration/export-setting/export-setting.component.html (hunk 15 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/qbd/configuration/export-setting/export-setting.component.html:50]: if i change the path it is not working

File: src/styles.scss (hunk 21 above)
Comment (DhaaraniCIT) [reply src/styles.scss:49]: it is not affecting bamboohr

File: src/app/shared/components/qbd/configuration/configuration-radio-field/configuration-radio-field.component.html (hunk 10 above)
Comment (DhaaraniCIT) [reply src/app/shared/components/qbd/configuration/configuration-radio-field/configuration-radio-field.component.html:2]: the width is 60%, so I used this

File: src/app/integrations/integrations.module.ts (hunk 5 above)
Comment (ashwin1111) [reply src/app/integrations/integrations.module.ts:21]: qbd component falls under qbd module no? why not import shared module there?

File: src/app/shared/components/qbd/configuration/export-setting/export-setting.component.html (hunk 15 above)
Comment (ashwin1111) [reply src/app/shared/components/qbd/configuration/export-setting/export-setting.component.html]: use <svg-icon-sprite> for icons

File: src/app/integrations/integrations.module.ts (hunk 5 above)
Comment (DhaaraniCIT) [reply src/app/integrations/integrations.module.ts:21]: it is imported in qbd, but still, it is not working

Comment (ashwin1111) [reply src/app/integrations/integrations.module.ts:21]: let's check on call in next sync up

**Comment by DhaaraniCIT** [comment]
https://user-images.githubusercontent.com/55541808/216962938-a581809e-98c8-45ed-b9ad-9ac7b7fa214f.mov
